### PR TITLE
Clean up docs to with regards to changes in SELFDESTRUCT opcode

### LINKED ISSUE #13889: Clean up documentation with regards to deprecated `selfdestruct`
As `selfdestruct` is to be deprecated, and now even triggers a deprecation warning (see https://github.com/ethereum/solidity/issues/13844, https://github.com/ethereum/solidity/pull/13884), the docs should be fully updated to reflect this, whether by adding additional deprecation warnings, or by altering any existing examples so as not to suggest the usage of `selfdestruct`.

Changes should be made in the following files:
```
docs/cheatsheet.rst
docs/contracts/function-modifiers.rst
docs/contracts/functions.rst
docs/contracts/inheritance.rst
docs/examples/micropayment.rst
docs/internals/optimizer.rst
docs/metadata.rst
docs/security-considerations.rst
docs/smtchecker.rst
docs/style-guide.rst
docs/using-the-compiler.rst
```

### PR DESCRIPTION
**Pull request by mehtavishwa30**
Clean up docs to reflect changes in the behaviour of `SELFDESTRUCT` as per [EIP-6780](https://eips.ethereum.org/EIPS/eip-6780).

Fixes #13889 

Checklist for docs files to be changed:

- [x] docs/cheatsheet.rst
- [x] docs/contracts/function-modifiers.rst
- [x] docs/contracts/functions.rst 
    -  no changes seem to be needed regarding the mention of `selfdestruct` there.
- [x] docs/contracts/inheritance.rst
- [x] docs/examples/micropayment.rst
- [x] docs/internals/optimizer.rst
    -  no changes seem to be needed regarding the mention of `selfdestruct` there.
- [x] docs/metadata.rst
- [x] docs/security-considerations.rst
    -  no changes seem to be needed regarding the mention of `selfdestruct` there.
- [x] docs/smtchecker.rst 
    -  no changes seem to be needed regarding the mention of `selfdestruct` there. However, we may need to account for the semantic change of `selfdestruct` in our tests when Cancun becomes the default EVM version.
- [x] docs/style-guide.rst
- [x] docs/using-the-compiler.rst

### CODE REVIEW

**Comment by cameel** [comment]
This looks like it's still work in progress? You should mark it as a draft if you're still working on it :)

**Comment by cameel** [comment]
Also, please add `Fixes #<issue number` (without backticks of course) in the PR description comment when a PR is meant to completely cover an issue. Github will then close the issue automatically when we merge the PR.

File: docs/cheatsheet.rst [hunk 1]
```diff
@@ -118,7 +118,7 @@ Contract-related
 
 - ``this`` (current contract's type): the current contract, explicitly convertible to ``address`` or ``address payable``
 - ``super``: a contract one level higher in the inheritance hierarchy
-- ``selfdestruct(address payable recipient)``: destroy the current contract, sending its funds to the given address
+- ``selfdestruct(address payable recipient)``: starting from ``EVM >= Cancun`` it will **only** sends all Ether in the account to the given recipient, except when ``selfdestruct`` is called in the same transaction the contract was created in; in that case, the behavior pre-Cancun hardfork is preserved (i.e., ``EVM <= Shanghai``), and ``selfdestruct`` will destroy the current contract, deleting any data, including storage keys, code, or the account itself.
```
Comment (matheusaaguiar) [inline docs/cheatsheet.rst]:
```suggestion
- ``selfdestruct(address payable recipient)``: starting from ``EVM >= Cancun`` it will **only** send all Ether in the account to the given recipient, except when ``selfdestruct`` is called in the same transaction the contract was created in; in that case, the behavior pre-Cancun hardfork is preserved (i.e., ``EVM <= Shanghai``), and ``selfdestruct`` will destroy the current contract, deleting any data, including storage keys, code and the account itself.
```

File: docs/contracts/inheritance.rst [hunk 2]
```diff
@@ -108,93 +115,170 @@ Details are given in the following example.
...
-        function destroy() public override(Base1, Base2) { super.destroy(); }
+        event Paused_Final();
+        function pause() public override(Base1, Base2) {
+            /* Here, we emit an event to simulate some Final pause logic */
+            emit Paused_Final();
+            super.pause();
+        }
     }
 
 If ``Base2`` calls a function of ``super``, it does not simply
```
Comment (matheusaaguiar) [inline docs/contracts/inheritance.rst]: Not strictly related to this PR, but shouldn't this be referring to `Final` instead of `Base2`?
```suggestion
If ``Final`` calls a function of ``super``, it does not simply
```

File: docs/examples/micropayment.rst [hunk 3]
```diff
@@ -80,6 +80,10 @@ the contract's address itself will be accepted. You can find
 an example of this in the first two lines of the ``claimPayment()``
 function of the full contract at the end of this section.
 
+Furthermore, instead of destroying the contract by calling ``selfdestruct``,
+which is currently deprecated, we will instead disable the contract's functionalities by freezing it,
+resulting in the reverting of any call after being frozen.
```
Comment (matheusaaguiar) [inline docs/examples/micropayment.rst]:
```suggestion
which is currently deprecated, we will disable the contract's functionalities by freezing it,
resulting in the reversion of any call after it being frozen.
```

File: docs/examples/micropayment.rst [hunk 4]
```diff
@@ -144,30 +148,54 @@ The full contract
...
-        address owner = msg.sender;
 
+    contract Owned {
+        address payable owner;
+        constructor() {
+            owner = payable(msg.sender);
+        }
+    }
+
+    contract Frozeable is Owned {
```
Comment (matheusaaguiar) [inline docs/examples/micropayment.rst]:
```suggestion
    contract Freezable is Owned {
```

File: docs/examples/micropayment.rst [hunk 5]
```diff
@@ -144,30 +148,54 @@ The full contract
...
+            _;
+        }
+
+        function freeze() internal {
+            if (msg.sender == owner)
+                _frozen = true;
+        }
+    }
+
+    contract ReceiverPays is Frozeable {
```
Comment (matheusaaguiar) [inline docs/examples/micropayment.rst]:
```suggestion
    contract ReceiverPays is Freezable {
```

File: docs/examples/micropayment.rst [hunk 6]
```diff
@@ -305,7 +332,7 @@ Closing the Payment Channel
 When Bob is ready to receive his funds, it is time to
 close the payment channel by calling a ``close`` function on the smart contract.
 Closing the channel pays the recipient the Ether they are owed and
-destroys the contract, sending any remaining Ether back to Alice. To
+deactivate the contract by freezing it, sending any remaining Ether back to Alice. To
```
Comment (matheusaaguiar) [inline docs/examples/micropayment.rst]:
```suggestion
deactivates the contract by freezing it, sending any remaining Ether back to Alice. To
```

File: docs/contracts/inheritance.rst (hunk 2 above)
Comment (r0qs) [reply docs/contracts/inheritance.rst]: Indeed.

File: docs/cheatsheet.rst [hunk 7]
```diff
@@ -118,7 +118,7 @@ Contract-related
 
 - ``this`` (current contract's type): the current contract, explicitly convertible to ``address`` or ``address payable``
 - ``super``: a contract one level higher in the inheritance hierarchy
-- ``selfdestruct(address payable recipient)``: destroy the current contract, sending its funds to the given address
+- ``selfdestruct(address payable recipient)``: starting from ``EVM >= Cancun`` it will **only** send all Ether in the account to the given recipient, except when ``selfdestruct`` is called in the same transaction the contract was created in; in that case, the behavior pre-Cancun hardfork is preserved (i.e., ``EVM <= Shanghai``), and ``selfdestruct`` will destroy the current contract, deleting any data, including storage keys, code and the account itself.
```
Comment (cameel) [inline docs/cheatsheet.rst]: This extra info should go under [Units and Globally Available Variables > Contract-related](https://docs.soliditylang.org/en/latest/units-and-global-variables.html#contract-related) as a note (or better a warning). Cheatsheet is supposed to be, well, a concise cheatsheet, so it should only have a very brief summary of the most important info that's already available in the full form elsewhere.

Also, [Introduction to Smart Contracts > Deactivate and Self-destruct](https://docs.soliditylang.org/en/latest/introduction-to-smart-contracts.html#deactivate-and-self-destruct) seems to be the place where `selfdestruct` is documented in more detail and this info should also be there.

Comment (cameel) [reply docs/cheatsheet.rst]: Also we should probably emphasize that this change in behavior depends on the EVM version of the chain where the contract is deployed and not the value passed to the compiler using the `--evm-version` flag. Many users might not realize the distinction.

File: docs/contracts/function-modifiers.rst [hunk 8]
```diff
@@ -41,16 +40,6 @@ if they are marked ``virtual``. For details, please see
         }
     }
 
-    contract destructible is owned {
-        // This contract inherits the `onlyOwner` modifier from
-        // `owned` and applies it to the `destroy` function, which
-        // causes that calls to `destroy` only have an effect if
-        // they are made by the stored owner.
-        function destroy() public onlyOwner {
```
Comment (cameel) [inline docs/contracts/function-modifiers.rst:49]: What this comment says about modifier inheritance is still relevant. You could put it over `Register.changePrice()` below and it would still make sense.

File: docs/contracts/inheritance.rst [hunk 9]
```diff
@@ -40,26 +40,35 @@ Details are given in the following example.
...
     // contracts can access all non-private members including
     // internal functions and state variables. These cannot be
     // accessed externally via `this`, though.
-    contract Destructible is Owned {
+    contract Pausable is Owned {
+        bool private paused;
+
+        constructor() {
+            paused = false;
+        }
```
Comment (cameel) [inline docs/contracts/inheritance.rst]: This could be simplified, also showing another variant of initialization:
```suggestion
        bool private paused = false;
```

Also the `&& !paused` and `&& paused` in the two conditions below seems superfluous because there are no side-effects other than the value of `paused` being changed.

File: docs/contracts/inheritance.rst [hunk 10]
```diff
@@ -108,93 +115,170 @@ Details are given in the following example.
...
+
+        constructor() { _paused = false; }
+
+        function paused() public view virtual returns (bool) {
+            return _paused;
+        }
+
+        function pause() virtual public {
+            if (msg.sender == owner && !_paused)
+            {
```
Comment (cameel) [inline docs/contracts/inheritance.rst]: This seems to be the prevalent style used for braces on this page and it would be best to keep it consistent:
```suggestion
            if (msg.sender == owner && !_paused) {
```

There are also more places like that below.

File: docs/contracts/inheritance.rst [hunk 11]
```diff
@@ -108,93 +115,170 @@ Details are given in the following example.
...
+                emit Unpaused();
+                _paused = false;
+            }
         }
     }
 
-    contract Base1 is Destructible {
-        function destroy() public virtual override { /* do cleanup 1 */ Destructible.destroy(); }
+    contract Base1 is Pausable {
+        event Paused_Base1();
```
Comment (cameel) [inline docs/contracts/inheritance.rst]:
```suggestion
        event Base1Paused();
```

File: docs/contracts/inheritance.rst [hunk 12]
```diff
@@ -108,93 +115,170 @@ Details are given in the following example.
...
+        }
+
+        function unpause() virtual public {
+            if (msg.sender == owner && _paused)
+            {
+                emit Unpaused();
+                _paused = false;
+            }
         }
     }
```
Comment (cameel) [inline docs/contracts/inheritance.rst]: This was meant to illustrate the inheritance and I think that the new example being this much larger makes this section lose focus a bit. IMO it would be better to still keep it concise. How about we drop `unpause()`, `paused()` and `_paused` and just have a single function that emits the `Paused` event, to replace `destroy()`?

Same for a similar contract below.

File: docs/using-the-compiler.rst [hunk 13]
```diff
@@ -236,12 +236,12 @@ Input Description
...
           ]
         },
-        "destructible":
+        "incrementable":
         {
           // Optional: keccak256 hash of the source file
           "keccak256": "0x234...",
           // Required (unless "urls" is used): literal contents of the source file
-          "content": "contract destructible is owned { function shutdown() { if (msg.sender == owner) selfdestruct(owner); } }"
+          "content": "contract incrementable is owned { uint256 private x = 0; function increment(uint256 _x) public { if (msg.sender == owner) x = _x; } }"
```
Comment (cameel) [inline docs/using-the-compiler.rst]: Nothing truly wrong here except that from the name I expected to see the function to `++_x` and was confused for a moment why it does not do that :) Maybe this would be better:

```suggestion
          "content": "contract settable is owned { uint256 private x = 0; function set(uint256 _x) public { if (msg.sender == owner) x = _x; } }"
```

Same in `docs/metadata.rst`.

**Review by cameel — COMMENTED** [review]
Overall looks good, but there are a few things, that should be corrected:
- `selfdestruct()` details were added only in the cheatsheet and not in the spots where it's primiarily documented. Should be the other way around.
- Some inheritance examples could be shortened. The new ones are long and distract from the main point IMO.

Other than that it's mostly minor stuff. See comments below for details.

File: docs/contracts/inheritance.rst (hunk 12 above)
Comment (r0qs) [reply docs/contracts/inheritance.rst]: But I think we should at least keep the `_paused` variable so that the contract can be "paused," for the sake of completeness in the example

File: docs/using-the-compiler.rst (hunk 13 above)
Comment (r0qs) [reply docs/using-the-compiler.rst]: ops :0

File: docs/contracts/inheritance.rst (hunk 12 above)
Comment (r0qs) [reply docs/contracts/inheritance.rst]: Actually, rethinking about it, you are completely right. I changed the example ;)

Comment (cameel) [reply docs/contracts/inheritance.rst]: I mean, I'm talking only about these short contracts that illustrate inheritance. Those were meant to be short and there's not much point including other distractions there.

But there's still a bigger pausable example above them and there having all the elements there is perfectly fine and even recommended. `Pausable` is a very good idea overall and I like it, just not specifically at the point I'm commenting on.

File: docs/introduction-to-smart-contracts.rst [hunk 14]
```diff
@@ -563,7 +563,9 @@ idea, but it is potentially dangerous, as if someone sends Ether to removed
 contracts, the Ether is forever lost.
 
 .. warning::
-    From version 0.8.18 and up, the use of ``selfdestruct`` in both Solidity and Yul will trigger a
+    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient. In the case when ``selfdestruct`` is called in the same transaction the contract was created in, the behaviour of ``selfdestruct`` described above the warning pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself.
```
Comment (r0qs) [inline docs/introduction-to-smart-contracts.rst]: What about:
```suggestion
    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract. However, when ``selfdestruct`` is called in the same transaction as the contract calling ``selfdestruct`` was created, the behaviour of ``selfdestruct`` pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself. See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.
```

File: docs/introduction-to-smart-contracts.rst [hunk 15]
```diff
@@ -563,7 +563,9 @@ idea, but it is potentially dangerous, as if someone sends Ether to removed
...
+
+    Also, from version 0.8.18 and up, the use of ``selfdestruct`` in both Solidity and Yul will trigger a
     deprecation warning, since the ``SELFDESTRUCT`` opcode will eventually undergo breaking changes in behavior
     as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
```
Comment (r0qs) [inline docs/introduction-to-smart-contracts.rst]: Maybe rephrase this to something like:
```suggestion
    Also, note that the ``selfdestruct`` opcode remains deprecated since Solidity version 0.8.18, and the use of the opcode in both Solidity and Yul will trigger a deprecation warning, since the opcode is expected to undergo substantial behavioral changes, as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
    Moreover, future adjustments to the EVM may further restrict the functionality of this opcode.
```

Comment (mehtavishwa30) [reply docs/introduction-to-smart-contracts.rst]: Yes, this makes more sense. Thank you! :)

File: docs/cheatsheet.rst (hunk 7 above)
Comment (mehtavishwa30) [reply docs/cheatsheet.rst]: @cameel just finished making changes to accommodate your feedback. Feel free to take a look if you can. :)

File: docs/cheatsheet.rst [hunk 16]
```diff
@@ -118,7 +118,7 @@ Contract-related
 
 - ``this`` (current contract's type): the current contract, explicitly convertible to ``address`` or ``address payable``
 - ``super``: a contract one level higher in the inheritance hierarchy
-- ``selfdestruct(address payable recipient)``: destroy the current contract, sending its funds to the given address
+- ``selfdestruct(address payable recipient)``: Starting from ``EVM >= Cancun``, ``selfdestruct`` will only transfer all Ether to the recipient, except when invoked in the same transaction that created the contract, in which case it preserves pre-Cancun behaviour.
```
Comment (cameel) [inline docs/cheatsheet.rst]: Well, now it does no longer says what the behavior is, only how it changes :)

IMO something like this would be more fitting for a cheat sheet:

```suggestion
- ``selfdestruct(address payable recipient)``: send all funds to the given address and (only on EVMs before Cancun or when invoked within the transaction creating the contract) destroy the contract.
```

File: docs/metadata.rst [hunk 17]
```diff
@@ -175,9 +175,10 @@ explanatory purposes.
       },
       // Required: Compilation source files/source units, keys are file paths
       "sources": {
-        "destructible": {
+        "settable": {
           // Required (unless "url" is used): literal contents of the source file
-          "content": "contract destructible is owned { function destroy() { if (msg.sender == owner) selfdestruct(owner); } }",
+          "content": "contract settable is owned { uint256 private x = 0; function set(uint256 _x) public { if (msg.sender == owner) x = _x; } }",
+
```
Comment (cameel) [inline docs/metadata.rst]: Stray empty line.

File: docs/introduction-to-smart-contracts.rst [hunk 18]
```diff
@@ -563,9 +563,10 @@ idea, but it is potentially dangerous, as if someone sends Ether to removed
...
-    deprecation warning, since the ``SELFDESTRUCT`` opcode will eventually undergo breaking changes in behavior
-    as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
+    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract. However, when ``selfdestruct`` is called in the same transaction as the contract calling ``selfdestruct`` was created, the behaviour of ``selfdestruct`` pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself. See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.
+
+    Also, note that the ``selfdestruct`` opcode remains deprecated since Solidity version 0.8.18, and the use of the opcode in both Solidity and Yul will trigger a deprecation warning, since the opcode is expected to undergo substantial behavioral changes, as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
+    Moreover, future adjustments to the EVM may further restrict the functionality of this opcode.
```
Comment (cameel) [inline docs/introduction-to-smart-contracts.rst]: Just a general remark - note that the docs break long lines. This makes them generally easier to edit and to read in the source form. The line breaks disappear when the document is rendered.

I'd particularly recommend breaking the line after each sentence (or earlier if it's still too long), because this way you don't have to redo the line breaks in the whole paragraph if you change it. It also makes the diff more readable (only the changed sentence is highlighted). It also makes it much easier to suggest short corrections on github - see in particular my next comment - I had to include the whole paragraph, but it would be much clearer if could comment just on the part I'm correcting.

Here's how I'd do this for this bit:

```suggestion
    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract.
    However, when ``selfdestruct`` is called in the same transaction as the contract calling ``selfdestruct`` was created,
    the behaviour of ``selfdestruct`` pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract,
    deleting any data, including storage keys, code and the account itself.
    See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.

    Also, note that the ``selfdestruct`` opcode remains deprecated since Solidity version 0.8.18, and the use of the opcode
    in both Solidity and Yul will trigger a deprecation warning, since the opcode is expected to undergo substantial behavioral changes,
    as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
    Moreover, future adjustments to the EVM may further restrict the functionality of this opcode.
```

File: docs/introduction-to-smart-contracts.rst [hunk 19]
```diff
@@ -563,9 +563,10 @@ idea, but it is potentially dangerous, as if someone sends Ether to removed
 contracts, the Ether is forever lost.
 
 .. warning::
-    From version 0.8.18 and up, the use of ``selfdestruct`` in both Solidity and Yul will trigger a
-    deprecation warning, since the ``SELFDESTRUCT`` opcode will eventually undergo breaking changes in behavior
-    as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
+    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract. However, when ``selfdestruct`` is called in the same transaction as the contract calling ``selfdestruct`` was created, the behaviour of ``selfdestruct`` pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself. See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.
```
Comment (cameel) [inline docs/introduction-to-smart-contracts.rst]: This bit sounds a bit awkward:
```suggestion
    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract. However, when ``selfdestruct`` is called in the same transaction that creates the contract calling it, the behaviour of ``selfdestruct`` before Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself. See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.
```

File: docs/introduction-to-smart-contracts.rst (hunk 18 above)
Comment (cameel) [inline docs/introduction-to-smart-contracts.rst]: Well, it has already undergone substantial behavioral changes :)

How about using the wording from the blog post?
```suggestion
    Also, note that the ``selfdestruct`` opcode has been deprecated in Solidity version 0.8.18, as recommended by `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
    The deprecation is still in effect and the compiler will still emit warnings on its use.
    Any use in newly deployed contracts is strongly discouraged even if the new behavior is taken into account.
    Future changes to the EVM might further reduce the functionality of the opcode.
```

File: docs/introduction-to-smart-contracts.rst [hunk 20]
```diff
@@ -563,9 +563,10 @@ idea, but it is potentially dangerous, as if someone sends Ether to removed
...
+
```
Comment (cameel) [inline docs/introduction-to-smart-contracts.rst]: One thing that's still missing is https://github.com/ethereum/solidity/pull/14803#discussion_r1511340392. I.e. I think it's important to mention the distinction between EVM version used by the network and the EVM version we're compiling for.

We could use the wording based on the blog post:
```
This is a network-wide change that affected all contracts present on the Ethereum mainnet and testnets.
The ``--evm-version`` setting used when compiling the contract has no bearing on it.
```

File: docs/units-and-global-variables.rst [hunk 21]
```diff
@@ -339,7 +339,9 @@ Contract-related
 Furthermore, all functions of the current contract are callable directly including the current function.
 
 .. warning::
-    From version 0.8.18 and up, the use of ``selfdestruct`` in both Solidity and Yul will trigger a
+    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient. In the case when ``selfdestruct`` is called in the same transaction the contract was created in, the behaviour of ``selfdestruct`` described above the warning pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself.
+
+    Also, from version 0.8.18 and up, the use of ``selfdestruct`` in both Solidity and Yul will trigger a
     deprecation warning, since the ``SELFDESTRUCT`` opcode will eventually undergo breaking changes in behavior
     as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
```
Comment (cameel) [inline docs/units-and-global-variables.rst]: Most of the comments above apply here too.

**Review by cameel — COMMENTED** [review]
Some minor details are still missing and I'd also adjust wording in a few places.

I provided suggestions for most comments so it should all be easy to apply.

File: docs/introduction-to-smart-contracts.rst (hunk 18 above)
Comment (mehtavishwa30) [reply docs/introduction-to-smart-contracts.rst]: Thanks for the pointers, @cameel! I will keep this in mind.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `f47ddc0bfa1d12c91c4f62419e0a05150c5c9bba`:
```
Error: Trailing whitespace found:
docs/units-and-global-variables.rst:355:    The new behaviour of ``self-destruct`` is the result of a network-wide change that affects all 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `82332dbac04cc259292ea8a1fb3ee61715746a8b`:
```
Error: Trailing whitespace found:
docs/introduction-to-smart-contracts.rst:584:    The new behaviour of ``self-destruct`` is the result of a network-wide change that affects all 
docs/units-and-global-variables.rst:355:    The new behaviour of ``self-destruct`` is the result of a network-wide change that affects all 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `71cd39d50611d8673c268fc3bd233a066a7a941a`:
```
Error: Trailing whitespace found:
docs/introduction-to-smart-contracts.rst:584:    The new behaviour of ``self-destruct`` is the result of a network-wide change that affects all 

```
Please check that your changes are working as intended.

File: docs/introduction-to-smart-contracts.rst (hunk 20 above)
Comment (cameel) [reply docs/introduction-to-smart-contracts.rst]: I see you added this to one of the other warnings. I think it's related to the one here instead so I just pushed a commit that moved it here.

**Review by cameel — APPROVED** [review]
Looks good now. I already squashed the commits so we can merge.

File: docs/introduction-to-smart-contracts.rst [hunk 22]
```diff
@@ -563,9 +563,23 @@ idea, but it is potentially dangerous, as if someone sends Ether to removed
...
+    the Ethereum mainnet and testnets.
+    It is important to note that this change is dependent on the EVM version of the chain on which
+    the contract is deployed.
+    The ``--evm-version`` setting used when compiling the contract has no bearing on it.
+
+    Also, note that the ``selfdestruct`` opcode has been deprecated in Solidity version 0.8.18,
+    as recommended by `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
+    The deprecation is still in effect and the compiler will still emit warnings on its use.
+    Any use in newly deployed contracts is strongly discouraged even if the new behavior is taken into account.
+    Future changes to the EVM might further reduce the functionality of the opcode.
```
Comment (cameel) [inline docs/introduction-to-smart-contracts.rst:582]: @mehtavishwa30 Taking a final look at the rendered version, this whole section about `selfdestruct` actually strikes me as too advanced for "Introduction to Smart Contracts". Mentioning `selfdestruct` made sense when it was in common use, but now that it has all these warnings, it's no longer recommended for newbies to touch it. Maybe we should simplify it for new users?

We could delete everything from it and leave only a short paragraph saying that contracts cannot be deleted, only deactivated. We still have a copy of most of these warnings in "Units and Global Variables" so we could link there for anyone who wants the details. But in the intro we'd only be mentioning the fact that something called `selfdestruct` exists, but you should no be tempted to use it for deleting contracts since it's deprecated or something like that.

I mean, this is not necessary in this PR, I'm merging it as is, but could be done in a follow-up PR if you want to brush up the intro page a little.
